codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java

```diff
@@ -330,7 +331,10 @@ private static String decodeComponent(String s, int from, int toExcluded, Charse
 
             byteBuf.clear();
             do {
-                byteBuf.put(decodeHexByte(s, i, toExcluded));
+                if (i + 3 > toExcluded) {
+                    throw new IllegalArgumentException("unterminated escape sequence at index " + i + " of: " + s);
+                }
+                byteBuf.put(decodeHexDump(s, i + 1, 2));
```

Comment:
this will incur an intermediate `byte[]` array and copy operation. We consider creating a variant of `decodeHexDump` which can directly write into a `ByteBuf` ... something like:

```java
public static void decodHexDump(CharSequence hexDump, int start, int len, ByteBuf result) {
...
}
```